chatkit-core/src/main/kotlin/com/pusher/chatkit/subscription/SubscriptionListener.kt

```diff
@@ -0,0 +1,30 @@
+package com.pusher.chatkit.subscription
+
+import com.pusher.platform.SubscriptionListeners
+import elements.EosError
+import elements.Headers
+import elements.SubscriptionEvent
+
+/**
+ * Platform SubscriptionListeners, but as an interface.
+ *
+ * In future, consider pushing this interface down in to platform as an alternative to the
+ * SubscriptionListeners data-class-full-of-closures approach.
+ */
+internal interface SubscriptionListener<A> {
+    fun onEnd(error: EosError?)
+    fun onError(error: elements.Error)
+    fun onEvent(elementsEvent: SubscriptionEvent<A>)
+    fun onOpen()
+    fun onRetrying()
+    fun onSubscribe()
```

Comment:
sorry should have noticed it at the first read, but better late than never: could we order the callback in the order they may be called with also putting edge-cases last? i.e. onSubscribe, onOpen, onEvent, onRetrying, onError, onEnd

Reply:
I don't know why I stuck with alphabetical!

I am concerned that in the case of non-terminal errors, we get onError, followed by onRetrying and that I'm going to need to go sperlunking in the platform lib to elide the onError and pass the error to onRetrying instead of it's going to be a pain to stitch them together.

Reply:
> I don't know why I stuck with alphabetical!

Probably as that's the order in the platform lib. The refactoring hat's always needed!

> I am concerned that in the case of non-terminal errors, we get onError, followed by onRetrying

Yeah that's my concern as well.

> that I'm going to need to go sperlunking in the platform lib

🤣 https://en.wikipedia.org/wiki/spelunking I love the photo of the muddy section!